packages/extras/src/lib/components/Environment/Environment.svelte

```diff
@@ -0,0 +1,133 @@
+		if (!isCubeMap && format == 'hdr') return new RGBELoader()
+		return new TextureLoader()
+	}
+
+	const loadEnvironment = () => {
+		if (!renderer) throw new Error('Threlte renderer undefined')
+		const pmremGenerator = new PMREMGenerator(renderer)
+		DefaultLoadingManager.onLoad = function () {
+			pmremGenerator.dispose()
+		}
+		const loader: any = pickLoader()
+
+		if (isCubeMap) {
+			pmremGenerator.compileCubemapShader()
+			loader.setPath(path || '').load(files, (texture: any) => {
+				currentEnvMap = texture
+				texture.encoding = encoding || sRGBEncoding
+				currentProcessedEnv = pmremGenerator.fromCubemap(texture).texture
+				setEnvironment()
+			})
+		} else {
+			pmremGenerator.compileEquirectangularShader()
+			loader.load(`${path || ''}${files}`, (texture: any) => {
+				currentEnvMap = texture
+				texture.encoding = encoding || LinearEncoding
+				currentProcessedEnv = pmremGenerator.fromEquirectangular(texture).texture
+				setEnvironment()
+			})
+		}
+		currentFormat = format
+	}
+
+	const setEnvironment = () => {
+		scene.environment = currentProcessedEnv
+		if (isBackground) scene.background = currentProcessedEnv
+		invalidate()
+	}
+
+	const toggleGroundEnv = (
+		groundEnv: GroundProjectedEnv | undefined,
+		groundEnvProps: EnvironmentProperties['groundProjection'],
+		envMap: Texture
+	) => {
+		if (groundEnv && !groundEnvProps) {
+			scene.remove(currentGroundEnv)
+			currentGroundEnv = undefined
+			invalidate('Removing ground projected environment')
+		}
+		if (!groundEnv && groundEnvProps && envMap) {
+			currentGroundEnv = new GroundProjectedEnv(envMap)
+			currentGroundEnv.scale.setScalar(groundEnvProps.scalar)
+			currentGroundEnv.radius = groundEnvProps.radius
+			currentGroundEnv.height = groundEnvProps.height
+			scene.add(currentGroundEnv)
+			invalidate('Enabling ground projected environment')
+		}
+		if (groundEnv && groundEnvProps) {
+			currentGroundEnv.radius = groundEnvProps.radius
+			currentGroundEnv.height = groundEnvProps.height
+			invalidate('Updating ground projected environment properties')
+		}
+	}
+
+	$: toggleGroundEnv(currentGroundEnv, groundProjection, currentEnvMap)
+
+	$: {
+		// Reload environment on path change
+		if (envPath != previousEnvPath || format != currentFormat) {
+			if (currentProcessedEnv) {
+				currentProcessedEnv.dispose()
+			}
+			loadEnvironment()
+			previousEnvPath = envPath
+		}
+
+		if (!isBackground && scene.background) {
+			scene.background = null
+			invalidate('Removing Environment as scene.background')
+		}
+		if (isBackground && !scene.background && currentProcessedEnv) {
+			scene.background = currentProcessedEnv
+			invalidate('Adding Environment as scene.background')
+		}
+	}
+
+	onDestroy(() => {
+		scene.environment = null
+		scene.background = null
+		if (currentProcessedEnv) currentProcessedEnv.dispose()
+		groundProjection = undefined
+		invalidate('Environment destroyed')
+	})
+</script>
```

Comment:
Do you think it's a good idea to outsource the GroundProjectedEnv into its own component to separate the logic and make this component more readable?

```html
{#if groundProjection}
	<GroundProjectedEnv {groundProjection} />
{/if}
```

Reply:
I actually thought about it yesterday. GroundProjectedEnv as a child of <Environment/>.  It will make the component more readable but but since it's not something that will get used very often, is it worth adding yet another component to already growing docs?

Reply:
We don't have to expose it and frankly I wouldn't since it would then be a part of `@threlte/core`. I would just make it a separate component to clean it up and separate domains.

Reply:
Also check out [drei's implementation](https://github.com/pmndrs/drei/blob/master/src/core/Environment.tsx) for an inspiration on separation of concerns.